shared/components/data/themes-list-fetcher/index.jsx

```diff
@@ -49,76 +49,103 @@ let ThemesListFetcher = React.createClass( {
 		search: React.PropTypes.string,
 		tier: React.PropTypes.string,
 		onRealScroll: React.PropTypes.func,
-		onLastPage: React.PropTypes.func
-	},
-
-	getInitialState: function() {
-		return Object.assign( getThemesState(), { loading: true } );
+		onLastPage: React.PropTypes.func,
+
+		themes: React.PropTypes.array.isRequired,
+		lastPage: React.PropTypes.bool.isRequired,
+		loading: React.PropTypes.bool.isRequired,
+		query: React.PropTypes.func.isRequired,
+		fetchNextPage: React.PropTypes.func.isRequired,
+		incrementThemesPage: React.PropTypes.func.isRequired,
+		fetchThemes: React.PropTypes.func.isRequired,
+		fetchJetpackThemes: React.PropTypes.func.isRequired,
 	},
 
 	componentDidMount: function() {
-		ThemesListStore.on( 'change', this.onThemesChange );
-		if ( this.props.site || this.props.isMultisite ) {
-			this.queryThemes( this.props );
-		}
-	},
-
-	componentWillUnmount: function() {
-		ThemesListStore.off( 'change', this.onThemesChange );
+		this.refresh( this.props );
 	},
 
 	componentWillReceiveProps: function( nextProps ) {
-		const ignoreProps = [ 'children', 'onLastPage', 'site' ];
+		console.log( 'prop-search-2', nextProps.search );
+		const propKeys = [ 'search', 'tier' ];
 
-		if ( isEqual(
-			omit( this.props, ignoreProps ),
-			omit( nextProps, ignoreProps ) ) ) {
-			return;
+		if ( propKeys.some( key => this.props[ key ] !== nextProps[ key ] ) ) {
+			console.log( 'props changed' );
+			this.refresh( nextProps );
 		}
+	},
 
-		if ( nextProps.site || nextProps.isMultisite ) {
-			this.queryThemes( nextProps );
+	refresh: function( props ) {
+		if ( this.props.site || this.props.isMultisite ) {
+			this.queryThemes( props );
 		}
 	},
 
 	queryThemes: function( props ) {
-		const { onLastPage } = this.props;
-		this.onLastPage = onLastPage ? once( onLastPage ) : null;
-		queryThemes( props );
-	},
+		const {
+			onLastPage,
+			site,
+			search,
+			tier,
+
+			query,
+			fetchNextPage
+		} = props;
 
-	onThemesChange: function() {
-		this.setState( getThemesState() );
+		this.onLastPage = onLastPage ? once( onLastPage ) : null;
 
-		const { page } = ThemesListStore.getQueryParams();
-		const { loading, lastPage } = this.state;
+		query( {
+			search,
+			tier,
+			page: 0,
+			perPage: Constants.PER_PAGE,
+		} );
 
-		if ( page > 1 && ! loading && lastPage ) {
-			this.onLastPage && this.onLastPage();
-		}
+		fetchNextPage( site );
```

Comment:
I got kind of confused between the AC `fetchNextPage` and the fetcher's `fetchNextPage` function. Perhaps namespace/rename?